### PR TITLE
Label unknown/deprecated settings on startup

### PR DESCRIPTION
**Pull request by pmatseykanets**
## Issue: 

https://github.com/rancher/rancher/issues/43103
 
## Problem

When a deprecated setting is removed from the codebase the corresponding `Setting` resource is left behind in the cluster. We would like to clean up such orphaned setting resources on Rancher startup and potentially prevent creating custom setting resources via the webhook.
 
## Solution

On Rancher startup we list all the setting resources in the cluster and mark all unknown/unrecognized (e.g. deprecated) settings with a label. In a future release e.g. 2.9 we'll be removing them instead. This gives an opportunity to someone who used setting resources, e.g. for integration purposes, to transition to something else.
 
## Testing

Migrate from 2.7 to 2.8+, and run Rancher which should lead to some deprecated/removed and any custom settings being marked as unknown e.g. `kubeconfig-token-ttl-minutes`.

### Automated Testing

* Test types added/modified:
    * Unit

## QA Testing Considerations

* Deprecated settings left after the upgrade.
* Custom settings created in the local cluster.
 
### Regressions Considerations

N/A

Existing / newly added automated tests that provide evidence there are no regressions:
N/A

### CODE REVIEW

**Review by crobby — COMMENTED** [review]
Seems pretty reasonable to me.  Probably needs a few tests, but this looks like a good approach.

**Comment by crobby** [comment]
> @crobby FYI tests were updated to cover this change among other things.

Indeed.  Looks good to me.  Did you sort out the cache issue yet?

**Comment by pmatseykanets** [comment]
> Did you sort out the cache issue yet?

@crobby The settings cache is not available at that time yet as the dashboard registers itself pretty early.

File: pkg/controllers/dashboardapi/settings/settings.go
Comment (crobby) [reply pkg/controllers/dashboardapi/settings/settings.go:163]: maybe "cattle.io/unknownsetting"

File: pkg/controllers/dashboardapi/settings/settings.go
Comment (MbolotSuse) [inline pkg/controllers/dashboardapi/settings/settings.go:199]: Nit: In reality you are using this as a bool to determine if this has been called a single time, but you are storing/using an int which you update every time. Since this has no meaning past the first iteration, using a bool might make more sense here.

File: pkg/controllers/dashboardapi/settings/settings.go
Comment (MbolotSuse) [inline pkg/controllers/dashboardapi/settings/settings.go:197]: Nit: Not sure what value we get on declaring this as a var here, when we could instantiate when needed below.

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (MbolotSuse) [inline pkg/controllers/dashboardapi/settings/settings_test.go:297]: Nit: Generally speaking, you want to test the exported interface when possible since that's what other packages can consume. 

I'm marking this as a nit due to my above recommendation that this be an exported function, and because this is a new function, so we have been known to make exceptions when you aren't touching an existing function.

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (MbolotSuse) [inline pkg/controllers/dashboardapi/settings/settings_test.go:284]: Nit: It's a good practice to provide as realistic returns as possible. In this case, that would mean returning the GroupResource for the setting CRD (`management.cattle.io` and `settings`).

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (MbolotSuse) [inline pkg/controllers/dashboardapi/settings/settings_test.go:27]: Nit: Might be good to assert (in some way) that the settings which were in the map don't have the unknown label on them (maybe in teh asserts on the test cases, also assert that finalSetting.Labels doesn't contain the label).

**Review by MbolotSuse — COMMENTED** [review]
Couple of nits and one larger concern on cleaning up values in a "SetAll" function, but no major functionality concerns.

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (pmatseykanets) [reply pkg/controllers/dashboardapi/settings/settings_test.go:284]: Agreed.

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (pmatseykanets) [reply pkg/controllers/dashboardapi/settings/settings_test.go:27]: Agreed.

File: pkg/controllers/dashboardapi/settings/settings.go
Comment (pmatseykanets) [reply pkg/controllers/dashboardapi/settings/settings.go:197]: This captures error returned from `s.settings.Get()` and `s.settings.Update()`.

> when we could instantiate when needed below.
We couldn't as we are reusing the `setting`. Thus needs to be declared explicitly.

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (pmatseykanets) [reply pkg/controllers/dashboardapi/settings/settings_test.go:303]: > Generally speaking, you want to test the exported interface when possible since that's what other packages can consume.
Not if you want to test an essential behavior that is not part of the exported API as in this case.

File: pkg/controllers/dashboardapi/settings/settings.go
Comment (pmatseykanets) [reply pkg/controllers/dashboardapi/settings/settings.go:156]: I'm not sure if I agree. This was a deliberate decision to encapsulate and hide this in the concrete settings provider.
I don't see just yet what would be benefits of enlarging the surface of the provider interface to include this behavior (presumably by adding `CleanUknown` method).

Even from the naming standpoint `SetAll` (**set** is the operative word) can be read as make sure the settings store has **exactly** the settings that we defined and nothing else. In other words it can be seen as enforcing what was missing in the original implementation.

Comment (crobby) [reply pkg/controllers/dashboardapi/settings/settings.go:156]: I see Michael's point, but I think I prefer having the cleanupUnknownSettings call in here rather than exposing it and forcing something else to call that in addition to SetAll.  I don't think we'd ever want to SetAll without cleaning things up, right?

**Review by crobby — DISMISSED** [review]
I think this is good as-is.  We can continue the talk about where cleanupUnknownSettings is called if needed though.

File: pkg/controllers/dashboardapi/settings/settings.go
Comment (MbolotSuse) [reply pkg/controllers/dashboardapi/settings/settings.go:156]: > I don't see just yet what would be benefits of enlarging the surface of the provider interface to include this behavior (presumably by adding CleanUknown method).

The benefit is explicit control by the caller and clear semantics. The current code forces the caller to accept removed settings when using this method.

> Even from the naming standpoint SetAll (set is the operative word) can be read as make sure the settings store has exactly the settings that we defined and nothing else. In other words it can be seen as enforcing what was missing in the original implementation.

I don't agree on this part. At best, I can say that I think it's ambiguous. I definitely don't think that we can make a statement on the original intention of this method. The current comment also seems to be directly against this - it notes that it sets everything in the map, but makes no statement on settings not in the map.

> I don't think we'd ever want to SetAll without cleaning things up, right?

I think there's a world where the caller wants to set multiple settings that they control without function calls and so uses a method like this. That being said, no one uses this function for that purpose today so I don't think that we need to code for it.

To summarize my thoughts on this, I don't agree with the reasoning for this part of the change. That being said, I think we can go with this as long as we update the comment on both the struct implementation and on the interface to indicate that all settings not specified in the map will be dropped.

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (MbolotSuse) [reply pkg/controllers/dashboardapi/settings/settings_test.go:303]: I would argue that this is part of the exported API, since this is part of the behavior that you expect when you call `SetAll

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (MbolotSuse) [inline pkg/controllers/dashboardapi/settings/settings_test.go:306]: Nit: Since your test cases already use it, you may want to consider using the const here:

```suggestion
	assert.Equal(t, store["unknown"].Labels[unknownSettingLabelKey], "true")
```

File: pkg/controllers/dashboardapi/settings/settings.go
Comment (pmatseykanets) [reply pkg/controllers/dashboardapi/settings/settings.go:156]: > To summarize my thoughts on this, I don't agree with the reasoning for this part of the change.

It seems like a matter of difference of opinion. I'm not very strongly attached to the proposed solution but think that it addresses the original issue without "spilling" the semantics. I'm not convinced that we have enough evidence to support the idea that the caller will ever care about being able to choose whether to clean or not the unknown settings; thus making it premature engraving it in the interface IMHO.

That being said I 💯 agree and added comments describing the behavior of the provider implementation as well as stating in the interface that the behavior in regards to unknown settings is undefined and is up to a concrete implementation.

File: pkg/controllers/dashboardapi/settings/settings_test.go
Comment (pmatseykanets) [reply pkg/controllers/dashboardapi/settings/settings_test.go]: And in the end I would agree 😄. Testing a narrower aspect in isolation can be beneficial though. But at the same time I see no reason why we can't test it through the call of the exported method especially in this particular case. The test is updated.

File: pkg/controllers/dashboardapi/settings/settings.go
Comment (MbolotSuse) [inline pkg/controllers/dashboardapi/settings/settings.go:89]: Nit:
```suggestion
// NOTE: All settings not provided in settingsMap will be marked as unknown, and may be removed in the future.
```

File: pkg/settings/setting.go
Comment (MbolotSuse) [inline pkg/settings/setting.go:289]: NIt: I would move this to the "SetAll" function, since this behavior is particular to that function (Set and Get don't exhibit this behavior and I think it would be unreasonable for them to do so). I would also rephrase to something like:

```suggestion
// NOTE: The behavior for treating unknown settings is undefined.
// A provider may choose to remove settings which it has a record of, but are not provided in a new call to SetAll.
```

**Review by MbolotSuse — COMMENTED** [review]
Couple of notes:
- Hold off on merging, AFAIK we are still in code freeze
- Please squash the commits before merging
- Couple of nits on the comment phrasing, but up to you on if you use them.

File: pkg/settings/setting.go
Comment (pmatseykanets) [reply pkg/settings/setting.go]: Thanks for better phrasing, but I'm keeping on the interface itself. The provider may choose not to have such behavior or implement it in a different manner (e.g. store the names of provided settings and have a goroutine that proactively checks and does the cleaning). In other words the interface shouldn't prescribe it.

**Review by MbolotSuse — DISMISSED** [review]
Do squash commits and double check merge timeline, but LGTM.

**Review by crobby — COMMENTED** [review]
Needs squash, but lgtm

**Comment by pmatseykanets** [comment]
@crobby I was planning on doing "Squash and merge" but happy to explicitly squash if needed.

**Review by crobby — DISMISSED** [review]

**Review by MbolotSuse — APPROVED** [review]

**Review by crobby — APPROVED** [review]

**Pull request merged**
